Replace illegal characters in Kubernetes labels

jenrik opened the pull request:
<!--

Please check the following tips:
1. Avoid using force-push and commands that require it (such as `commit --amend` and `rebase origin/main`). This makes it more difficult for the maintainers to review your work. Add new commits on top of the current branch, and merge the new state of `main` into your branch with plain `merge`.
2. Provide a meaningful title for this pull request. It will be used as the commit message when this pull request is merged. Add as many commits as you like with any messages you like, they will be squashed into one commit.
3. If this pull request fixes an issue, refer to the issue with messages like `Closes #1234`, or `Fixes #1234` in the pull description.
4. Please check that you are targeting the `main` branch. Pull requests on release branches are only allowed for backports.
5. Make sure you have read contribution guidelines: https://woodpecker-ci.org/docs/development/getting-started
6. It is recommended to enable "Allow edits by maintainers", so maintainers can help you more easily.

-->

This is a workaround for Kubernetes labels being more restrictive that what we allow in Woodpecker.

xoxys commented:
Just reuse the logic from https://github.com/woodpecker-ci/woodpecker/blob/a9ea85742b2f4c146f7eaa0033c205f8eeb82616/pipeline/backend/kubernetes/pod.go#L132C33-L134?

jenrik commented:
That is a good idea. I have switched it over to use `toDNSName` for labels, and removed the sanitization of label keys because the `toDNSName` function removes `/` characters.

xoxys commented on pipeline/backend/kubernetes/pod.go:
```diff
@@ -108,7 +108,11 @@ func podLabels(step *types.Step, config *config, options BackendOptions) (map[st
 		// Internal labels are filtered on the server-side.
 		if config.PodLabelsAllowFromStep || strings.HasPrefix(k, pipeline.InternalLabelPrefix) {
-			labels[k] = v
+			sanitizedValue, err := toDNSName(v)
+			if err != nil {
+				return nil, err
+			}
+			labels[k] = sanitizedValue
```
```suggestion
			labels[k], err = toDNSName(v)
			if err != nil {
				return labels, err
			}
```

Can we not skip the extra spep here as well? I would also return `labels` on errors instead of `nil` to follow the existing implementation.

jenrik replied:
fixed in 41dd58c573284b7c7e2c43b71e45403e0892035a

xoxys replied:
Only partially :) The suggestion also drops the extra assignment `sanitizedValue, err := toDNSName(v)`

xoxys commented on pipeline/backend/kubernetes/secrets.go:
```diff
@@ -260,7 +260,11 @@ func registrySecretLabels(step *types.Step, config *config) (map[string]string,
 		// Internal labels are filtered on the server-side.
 		if config.PodLabelsAllowFromStep || strings.HasPrefix(k, pipeline.InternalLabelPrefix) {
-			labels[k] = v
+			sanitizedValue, err := toDNSName(v)
```
Same here.

xoxys commented:
Can we add a test case for this? We should prevent adding labels with illegal chars in the future.

jenrik commented:
We could probably pull the full regex from Kubernetes' source code and use that as part of a unit test

xoxys commented:
I'm totally fine to hardcode the test regex from the Kubernetes docs.
